cpp/include/cuspatial/experimental/detail/linestring_intersection_count.cuh

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/detail/utility/device_atomics.cuh>
+#include <cuspatial/detail/utility/linestring.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/tuple.h>
+
+namespace cuspatial {
+namespace detail {
+
+template <typename MultiLinestringRange1,
+          typename MultiLinestringRange2,
+          typename OutputIt1,
+          typename OutputIt2>
+__global__ void count_intersection_and_overlaps_simple(MultiLinestringRange1 multilinestrings1,
+                                                       MultiLinestringRange2 multilinestrings2,
+                                                       OutputIt1 point_count_it,
+                                                       OutputIt2 segment_count_it)
+{
+  using T          = typename MultiLinestringRange1::element_t;
+  using count_type = std::size_t;
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings1.num_points();
+       idx += gridDim.x * blockDim.x) {
+    auto const part_idx = multilinestrings1.part_idx_from_point_idx(idx);
+    if (!multilinestrings1.is_valid_segment_id(idx, part_idx)) continue;
+    auto const geometry_idx = multilinestrings1.geometry_idx_from_part_idx(part_idx);
+    auto [a, b]             = multilinestrings1.segment(idx);
+
+    for (auto const& linestring2 : multilinestrings2[geometry_idx]) {
+      for (auto [c, d] : linestring2) {
+        auto [point_opt, segment_opt] = segment_intersection(segment<T>{a, b}, segment<T>{c, d});
+        if (point_opt.has_value()) {
+          auto r = make_atomic_ref(*(point_count_it + geometry_idx));
+          r.fetch_add(1);
```

Comment:
Should look at the SASS generated for this. I know that if you use `atomicAdd` or `atomicInc` the compiler should automatically generate warp-aggregated atomics to reduce atomic memory collisions. But I'm not sure if this `atomic_ref ::fetch_add` reduces to the same code.

See https://developer.nvidia.com/blog/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/ (note update comment about automatically generated warp-aggregated atomics in the introduction)

Reply:
I think they are similar: https://godbolt.org/z/T6sshsdMz

Both instruction uses
```
 FLO.U32 R8, R6 
 POPC R9, R6 
 ISETP.EQ.U32.AND P0, PT, R8, R3, PT 
 MOV R3, c[0x0][0x16c]
```
warp aggregate. The difference is that `cuda::atomic_ref::fetch_add` has addtional `MEMBAR` and `CCTL` commands, not sure what they are for.